### PR TITLE
Check if input is countable before trying to count it

### LINKED ISSUE #44: "Parameter must be an array or an object that implements Countable"
I get this warning every time I use `dpm()`:

> Warning: count(): Parameter must be an array or an object that implements Countable in dprint_r() (line 1697 of /[SNIP]/modules/devel/devel.module).

Researched it and it's because in PHP v7.2.0 "`count()` will now yield a warning on invalid countable types passed to the `array_or_countable` parameter." ([source](https://www.php.net/manual/en/function.count.php#refsect1-function.count-changelog)).

Ideal solution would be to simply check [`is_countable()`](https://www.php.net/manual/en/function.is-countable.php) first, but it's only available in PHP >= v7.3.0. So instead, based on a comment on that page, I've made a PR that checks via a polyfill (and fixes the warnings).

### PR DESCRIPTION
**Pull request by ghost**
Fixes https://github.com/backdrop-contrib/devel/issues/44

### CODE REVIEW

**Comment by klonos** [comment]
...added the changes included in https://git.drupalcode.org/project/devel/commit/bd4061f

File: lib/krumo/class.krumo.php [hunk 1]
```diff
@@ -1099,7 +1099,7 @@
 ?>
 <li class="krumo-child">
 
-  <div class="krumo-element<?php echo count($data) > 0 ? ' krumo-expand' : '';?>"
+  <div class="krumo-element<?php echo ' krumo-expand';?>">
```
Comment (ghost) [inline lib/krumo/class.krumo.php]: The PHP tags here seem irrelevant now...

File: lib/krumo/class.krumo.php [hunk 2]
```diff
@@ -1110,7 +1110,7 @@
       <strong class="krumo-class"><?php echo get_class($data);?></strong>
   </div>
 
-  <?php if (count($data)) {
+  <?php {
```
Comment (ghost) [inline lib/krumo/class.krumo.php]: Should we remove the curly brackets and make this one line?

**Review by ghost — COMMENTED** [review]
See questions below re. some possible changes needed.

File: lib/krumo/class.krumo.php (hunk 1 above)
Comment (klonos) [reply lib/krumo/class.krumo.php]: Makes perfect sense 👍 ...done!

File: lib/krumo/class.krumo.php (hunk 2 above)
Comment (klonos) [reply lib/krumo/class.krumo.php]: ...also makes perfect sense. Also done 🙂

File: devel.module [hunk 3]
```diff
@@ -1737,9 +1737,11 @@ function ddebug_backtrace($return = FALSE, $pop = 0) {
       array_shift($backtrace);
     }
     $counter = count($backtrace);
```
Comment (ghost) [inline devel.module:1739]: Unlike the other `count()`'s, this one seems to have remain untouched...?

**Comment by klonos** [comment]
...we are missing some cross-ports. Please hold off reviewing/merging.

See: https://www.drupal.org/node/2855666 and the respective commit: https://git.drupalcode.org/project/devel/commit/752a1ca

**Comment by klonos** [comment]
This should be held off till we have all the crossports merged:

https://github.com/backdrop-contrib/devel/issues/49
https://github.com/backdrop-contrib/devel/issues/50

File: lib/krumo/class.krumo.php [hunk 4]
```diff
@@ -1047,7 +1047,7 @@
 ?>
 <li class="krumo-child">
 
-  <div class="krumo-element<?php echo count($data) > 0 ? ' krumo-expand' : '';?>"
+  <div class="krumo-element<?php echo !empty($data) ? ' krumo-expand' : '';?>">
     <?php if (count($data) > 0) {?> onClick="krumo.toggle(this);"<?php } ?>
     onMouseOver="krumo.over(this);"
```
Comment (alanmels) [inline lib/krumo/class.krumo.php:1052]: IMO, the whole block should look like:

```PHP
<div class="krumo-element<?php echo !empty($data) ? ' krumo-expand' : '';?>"
    <?php if (!empty($data)) {?> onClick="krumo.toggle(this);"<?php } ?>
    onMouseOver="krumo.over(this);"
    onMouseOut="krumo.out(this);">
```

Reasons:

1. The closing tag `>` on line 1050 you added will break the whole thing, because it duplicates the original closing in the very end of line 1053:

```PHP
onMouseOut="krumo.out(this);">
```

2. Since the lines 50-53 are to output a single HTML code, the same offending variable comparison `count($data) > 0` in line 1051 should be replaced with just `!empty($data)`.

File: lib/krumo/class.krumo.php [hunk 5]
```diff
@@ -1099,7 +1099,7 @@
 ?>
 <li class="krumo-child">
 
-  <div class="krumo-element<?php echo count($data) > 0 ? ' krumo-expand' : '';?>"
+  <div class="krumo-element krumo-expand">
     <?php if (count($data) > 0) {?> onClick="krumo.toggle(this);"<?php } ?>
     onMouseOver="krumo.over(this);"
     onMouseOut="krumo.out(this);">
```
Comment (alanmels) [inline lib/krumo/class.krumo.php:1105]: Consider changing with:

```PHP
<div class="krumo-element<?php echo !empty($data) ? ' krumo-expand' : '';?>"
    <?php if (!empty($data)) {?> onClick="krumo.toggle(this);"<?php } ?>
    onMouseOver="krumo.over(this);"
    onMouseOut="krumo.out(this);">
```

**Review by alanmels — CHANGES_REQUESTED** [review]
Also lines 1081, 1082 and 1083 should be replaced with:

```PHP
<?php  krumo::_vars($data);  ?>
```

in the same manner as you did on line 113.

File: lib/krumo/class.krumo.php (hunk 4 above)
Comment (klonos) [reply lib/krumo/class.krumo.php:1052]: Sorry for the late responce @alanmels 

You are correct. I have fixed this.

File: lib/krumo/class.krumo.php (hunk 5 above)
Comment (klonos) [reply lib/krumo/class.krumo.php:1105]: Right once again here. Also fixed 👍

File: devel.module (hunk 3 above)
Comment (klonos) [reply devel.module:1739]: @BWPanda I believe that the issue happens within conditionals, when trying to count something "uncountable". If you look at the code above, you'll see that `$backtrace` is actually set to `debug_backtrace()`, which according to PHP docs always returns an array (which is "countable"). So I believe we're safe here.

...besides, I would not want us to deviate too much from the 7.x commit here: https://git.drupalcode.org/project/devel/commit/bd4061f

Comment (klonos) [reply devel.module:1739]: ...if we are worried about every instance of `count()`, perhaps we should then introduce a `backdrop_count()` wrapper function, which will be providing an additional check for `instanceof Countable`. But that would be a separate issue, against core.

**Review by alanmels — COMMENTED** [review]
Now everything looks good to me.
